### PR TITLE
Function details full

### PR DESCRIPTION
**Pull request by woocash2**

### CODE REVIEW

**Comment by ggawryal** [comment]
This is a merged branch for both `generateCall` and `genRead/Write` tasks, with implementation from both sprints (previous code was not merged).
Note, that we've added some FunctionDetailsGenerator fields, so generating `FunctionDetailsGenerator` for all functions will require to be updated - temporarily, We've commented appropriate lines and disabled one test: `FunctionDependenciesAnalyzerTest.test function details generator creation`.

File: app/src/test/kotlin/compiler/intermediate_form/ControlFlowGraphBuilderTest.kt
Comment (bebidek) [inline app/src/test/kotlin/compiler/intermediate_form/ControlFlowGraphBuilderTest.kt:1]: Same as with ControlFlowGraphBuilder.kt

File: app/src/test/kotlin/compiler/intermediate_form/FunctionDetailsGeneratorTest.kt
Comment (bebidek) [inline app/src/test/kotlin/compiler/intermediate_form/FunctionDetailsGeneratorTest.kt:1]: I don't analyze this file in this review as some aspects of FDG class are to be discussed/corrected.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Good idea

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Good to know, removed

File: app/src/main/kotlin/compiler/intermediate_form/Addressing.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/Addressing.kt]: Yup, we've decided, that we'd only need `MemoryLabel` IFTNode. Removed this file totally and moved `memoryUnitSize` to `DefaultFunctionDetailsGenerator.kt`

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: On the other hand it documents what this function does in one place, so that reader wouldn't have to read the whole method. But moved as you suggested, since for not too long methods it might be cleaner.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Agreed & renamed

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Right, renamed it

File: app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt]: I also prefer `MemoryLabel` name, the other name might've been selected incorrectly by incautious merge conflict resolving.

File: app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt]: I forgot about that, let's remove it

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: It is tricky, because we'd have to watch out to not assign not already backed up callee-saved register to that virtual register storing the copy

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: True, I've moved saving into rax to `genEpilogue`. 
Although, we need `variableToStoreFunctionResult` and creating such temporary variable in `ControlFlow.createGraphForEachFunction`, because we need to store somewhere the result and `createGraphForEachFunction` shouldn't know anything about `Register` backend, so we'd rather not use `RAX` name there. Alternatively, we could pass the `Register` to that function to which it should write the result, but introducing a new `Variable` is cleaner imho, as it moves the responsibility to decide the actual Register (or memory address) for such variable to graph coloring algorithm we'll be using for allocating physical registers.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Thanks for the catch changed to copying and added removing them from stack to `genCall`

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: I see, moved to the end of that function, it should be safe to call these methods there

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Moved pushing them to stack explicitly after allocating space on stack for local variables, but if you think my previous comment isn't a problem, I'd be happy to employ your solution

Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Well, we're going to have some method of forbidding arbitrary register assignments (by adding extra edges to permanently colored vertices). We don't know yet how such an information will be passed there, so maybe (probably) some interfaces will have to be updated, but I'm pretty sure that it will be possible to do it in a clear way.

**Comment by bebidek** [comment]
One thing I recalled about SysV calling convention: they require the stack to be 16B aligned (i.e. RSP % 16 = 0) immediately before 'call'. Problem is it's not trivial to determine that because at this point we don't know how much temporary storage on stack will be used. And adding explicit runtime check before call seems like a horrible solution.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: OK, I guess I didn't understand the function of that variable well enough. As for the `Variable` vs `Register`, using a `Register` would also let the coloring algorithm decide (as long as we pass a virtual register, not one of the physical ones). But since we already have means to write the expression result to a variable, I agree that there's no need to change that.
One thing that bothers me is that we create new `Variable` object after all the semantic analysis phases, consequently making some maps incorrect (incomplete). It seems that it's not a problem now, but it has a potential to become a problem in the future.

Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Also, function can have multiple 'return' statements and we want to use the same variable in all of them, so we should create it only once.

File: app/src/main/kotlin/compiler/intermediate_form/Addressing.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/Addressing.kt]: It actually turns out that while we don't need it here, another PR (which adds assembly instruction set) uses this class (with some modifications), so the file is going to be reintroduced there. Still, it can be removed here.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Ok, updated it so it grows upwards now.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: I think it's pretty compact right now. If we introduced a method like `genAddress(base, offset)` then we'd still have to pass it the same arguments as before so it will kind of look the same.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: I've separated out some common logic

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: No, what I meant was to extract something like `getDisplaySlotAddress` (without arguments, just for display).

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Ok

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: I would suggest extracting even more, like that:
```
    private fun genAccess(namedNode: NamedNode, isDirect: Boolean, regAccessGenerator: (Register) -> IFTNode, memAccessGenerator: (IFTNode) -> IFTNode): IFTNode {
        return if (isDirect) {
            when (variablesLocationTypes[namedNode]!!) {
                VariableLocationType.MEMORY -> {
                    memAccessGenerator(
                        IntermediateFormTreeNode.Subtract(
                            IntermediateFormTreeNode.RegisterRead(Register.RBP),
                            IntermediateFormTreeNode.Const(variablesStackOffsets[namedNode]!!.toLong())
                        )
                    )
                }

                VariableLocationType.REGISTER ->
                    regAccessGenerator(variablesRegisters[namedNode]!!)
            }
        } else {
            if (variablesLocationTypes[namedNode]!! == VariableLocationType.REGISTER)
                throw IndirectRegisterAccess()

            memAccessGenerator(
                IntermediateFormTreeNode.Subtract(
                    IntermediateFormTreeNode.MemoryRead(displayElementAddress()),
                    IntermediateFormTreeNode.Const(variablesStackOffsets[namedNode]!!.toLong())
                )
            )
        }
    }

    override fun genRead(namedNode: NamedNode, isDirect: Boolean): IntermediateFormTreeNode =
        genAccess(namedNode, isDirect, {IntermediateFormTreeNode.RegisterRead(it)}, {IntermediateFormTreeNode.MemoryRead(it)})

    override fun genWrite(namedNode: NamedNode, value: IntermediateFormTreeNode, isDirect: Boolean): IntermediateFormTreeNode =
        genAccess(namedNode, isDirect, {IntermediateFormTreeNode.RegisterWrite(it, value)}, {IntermediateFormTreeNode.MemoryWrite(it, value)})

```
I used some abbreviations, but I think the idea is clear.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Moved variable creation to the beginning of this file to make sure it is done only once. For semantic analysis, we can create a separate issue for it, maybe linked with passing additional parameters to creating `DefaultFunctionDetailsGenerator` in `FunctionDependenciesAnalyzer.createFunctionDetailsGenerators`, because we have to pass that variable to the constructor of `DFDG` somehow anyway.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt]: Agree, applied refactor you suggested

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt]: Resolved some merge conflicts with that PR, trying to resolve conflicts in favour of that branch as much as possible, but please look at this file to make sure I haven't rolled back some refactors.

File: app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt
Comment (bebidek) [inline app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt:2]: We don't have tests for `genPrologue`/`genEpilogue`, but these would probably be ugly tests.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt]: Reverted `Links` name

File: app/src/main/kotlin/compiler/intermediate_form/DefaultFunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/DefaultFunctionDetailsGenerator.kt]: Great catch, updated

File: app/src/main/kotlin/compiler/intermediate_form/DefaultFunctionDetailsGenerator.kt
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate_form/DefaultFunctionDetailsGenerator.kt]: Indeed, moved it before restoring that register value

File: app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt
Comment (ggawryal) [reply app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt:2]: I can't see any way to test it without repeating exactly the same code in these tests as it is in the `genPrologue/Epilogue`, so it makes no sense to me. Maybe we could check if the order of pushing/popping items to stack is consistent between prologue and epilogue, but it is still only a fraction of the responsibility of these methods. The only method of testing it that comes to my mind are some integration tests later, when we'll have full translation to assembler - we'd compile it, call, and assert there were no runtime errors in asm code.

**Comment by ggawryal** [comment]
> One thing I recalled about SysV calling convention: they require the stack to be 16B aligned (i.e. RSP % 16 = 0) immediately before 'call'. Problem is it's not trivial to determine that because at this point we don't know how much temporary storage on stack will be used. And adding explicit runtime check before call seems like a horrible solution. Maybe we can forget about it for now (adding some TODO note). I think it doesn't make a difference as long as we don't call external functions (which can assume that) or deal with floating-point registers (some operations require aligned addresses).

We'd probably have some printf/scanfs, either as syscalls or linked from C lib, so eventually not alligned stack will be a problem

File: app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt
Comment (bebidek) [reply app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt:2]: Sounds reasonable.

**Review by bebidek — APPROVED** [review]

File: app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt
Comment (bebidek) [inline app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt:202]: I noticed this `println` which I believe doesn't belong here

**Comment by bebidek** [comment]
Also, the file 'DefaultDefaultFunctionDetailsGeneratorTest' is has a mistake in its name. I gave approve, so after fixing these two issues you can merge.

File: app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt
Comment (ggawryal) [reply app/src/test/kotlin/compiler/intermediate_form/DefaultDefaultFunctionDetailsGeneratorTest.kt]: Also the file had duplicated `Default` in name, fixed both

**Pull request merged**
